File: electron-app/bundler.py
#!/usr/bin/env python3
"""Bundler for Electron applications.

This script takes an Electron binary and app files, and bundles them
into a tar archive that can be extracted and run.
"""

import sys
import os
import zipfile
import tarfile
import tempfile
import shutil
import json
from pathlib import Path


def extract_electron(electron_zip_path, extract_dir):
    """Extract the Electron binary from the zip file, preserving symlinks."""
    print(f"Extracting Electron from {electron_zip_path}")

    # Use system unzip command to preserve symlinks
    # Python's zipfile module doesn't handle symlinks properly
    import subprocess
    result = subprocess.run(['unzip', '-q', electron_zip_path, '-d', extract_dir],
                          capture_output=True, text=True)
    if result.returncode != 0:
        raise Exception(f"Failed to extract Electron: {result.stderr}")

    print(f"Extracted to {extract_dir}")


def create_app_bundle(temp_dir, app_name, main_js_path, index_html_path, asset_paths):
    """Create the app bundle with the provided files."""
    # Find the Electron.app in the extracted files
    electron_app_path = None
    for root, dirs, files in os.walk(temp_dir):
        if 'Electron.app' in dirs:
            electron_app_path = os.path.join(root, 'Electron.app')
            break
    
    if not electron_app_path:
        raise Exception("Could not find Electron.app in extracted files")
    
    # Rename Electron.app to the app name
    app_bundle_path = os.path.join(temp_dir, f"{app_name}.app")
    shutil.move(electron_app_path, app_bundle_path)
    
    # Create the app directory inside the bundle
    app_resources_dir = os.path.join(app_bundle_path, 'Contents', 'Resources', 'app')
    os.makedirs(app_resources_dir, exist_ok=True)
    
    # Copy main.js
    shutil.copy2(main_js_path, os.path.join(app_resources_dir, 'main.js'))
    
    # Copy index.html
    shutil.copy2(index_html_path, os.path.join(app_resources_dir, 'index.html'))
    
    # Copy additional assets
    for asset_path in asset_paths:
        if os.path.isfile(asset_path):
            shutil.copy2(asset_path, app_resources_dir)
        elif os.path.isdir(asset_path):
            # Copy directory contents
            dest_dir = os.path.join(app_resources_dir, os.path.basename(asset_path))
            shutil.copytree(asset_path, dest_dir)
    
    # Create package.json
    package_json = {
        "name": app_name,
        "version": "1.0.0",
        "main": "main.js"
    }
    with open(os.path.join(app_resources_dir, 'package.json'), 'w') as f:
        json.dump(package_json, f, indent=2)
    
    # Update Info.plist to use the app name
    info_plist_path = os.path.join(app_bundle_path, 'Contents', 'Info.plist')
    if os.path.exists(info_plist_path):
        with open(info_plist_path, 'r') as f:
            content = f.read()
        # Replace Electron with app name in key places
        content = content.replace('<string>Electron</string>', f'<string>{app_name}</string>', 1)
        content = content.replace('<string>com.github.Electron</string>', f'<string>com.gazel.{app_name}</string>')
        with open(info_plist_path, 'w') as f:
            f.write(content)
    
    return app_bundle_path


def create_tar_archive(app_bundle_path, output_tar_path):
    """Create a tar archive of the app bundle."""
    print(f"Creating tar archive at {output_tar_path}")

    # First, fix permissions on the executable
    executable_path = os.path.join(app_bundle_path, 'Contents', 'MacOS', 'Electron')
    if os.path.exists(executable_path):
        os.chmod(executable_path, 0o755)
        print(f"Set execute permissions on {executable_path}")

    # Fix permissions on the main framework
    framework_path = os.path.join(app_bundle_path, 'Contents', 'Frameworks',
                                  'Electron Framework.framework', 'Versions', 'A',
                                  'Electron Framework')
    if os.path.exists(framework_path):
        os.chmod(framework_path, 0o755)
        print(f"Set execute permissions on {framework_path}")

    # Also fix permissions on helper executables
    helpers_dir = os.path.join(app_bundle_path, 'Contents', 'Frameworks',
                              'Electron Framework.framework', 'Versions', 'A', 'Helpers')
    if os.path.exists(helpers_dir):
        for helper in os.listdir(helpers_dir):
            helper_path = os.path.join(helpers_dir, helper)
            if os.path.isfile(helper_path) and not os.path.islink(helper_path):
                os.chmod(helper_path, 0o755)
                print(f"Set execute permissions on {helper_path}")

    # Fix permissions on libraries
    libs_dir = os.path.join(app_bundle_path, 'Contents', 'Frameworks',
                           'Electron Framework.framework', 'Versions', 'A', 'Libraries')
    if os.path.exists(libs_dir):
        for lib in os.listdir(libs_dir):
            if lib.endswith('.dylib'):
                lib_path = os.path.join(libs_dir, lib)
                if os.path.isfile(lib_path) and not os.path.islink(lib_path):
                    os.chmod(lib_path, 0o755)
                    print(f"Set execute permissions on {lib_path}")

    # Create tar with proper symlink handling
    with tarfile.open(output_tar_path, 'w') as tar:
        # Add the app bundle to the tar, preserving symlinks
        app_name = os.path.basename(app_bundle_path)
        tar.add(app_bundle_path, arcname=app_name, filter=None)

    print(f"Created tar archive: {output_tar_path}")


def main():
    if len(sys.argv) < 6:
        print("Usage: bundler.py <output_tar> <app_name> <main_js> <index_html> <electron_zip> [assets...]")
        sys.exit(1)
    
    output_tar = sys.argv[1]
    app_name = sys.argv[2]
    main_js_path = sys.argv[3]
    index_html_path = sys.argv[4]
    electron_zip_path = sys.argv[5]
    asset_paths = sys.argv[6:] if len(sys.argv) > 6 else []
    
    print(f"Building Electron app: {app_name}")
    print(f"Output: {output_tar}")
    print(f"Main JS: {main_js_path}")
    print(f"Index HTML: {index_html_path}")
    print(f"Electron ZIP: {electron_zip_path}")
    print(f"Assets: {asset_paths}")
    
    # Create a temporary directory for extraction
    with tempfile.TemporaryDirectory() as temp_dir:
        # Extract Electron
        extract_electron(electron_zip_path, temp_dir)
        
        # Create the app bundle
        app_bundle_path = create_app_bundle(
            temp_dir, app_name, main_js_path, index_html_path, asset_paths
        )
        
        # Create the tar archive
        create_tar_archive(app_bundle_path, output_tar)
    
    print(f"Successfully created {app_name} bundle")


if __name__ == "__main__":
    main()

File: electron/bundler.js
#!/usr/bin/env node
/**
 * Bundler for Electron applications.
 * 
 * This script takes an Electron binary and app files, and bundles them
 * into a tar archive that can be extracted and run.
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');
const tar = require('tar');
const os = require('os');

/**
 * Extract the Electron binary from the zip file, preserving symlinks.
 */
function extractElectron(electronZipPath, extractDir) {
    console.log(`Extracting Electron from ${electronZipPath}`);
    
    // Use system unzip command to preserve symlinks
    // Node's built-in zip modules don't handle symlinks properly
    try {
        execSync(`unzip -q "${electronZipPath}" -d "${extractDir}"`, {
            stdio: 'pipe'
        });
    } catch (error) {
        throw new Error(`Failed to extract Electron: ${error.message}`);
    }
    
    console.log(`Extracted to ${extractDir}`);
}

/**
 * Find Electron.app in the extracted files
 */
function findElectronApp(dir) {
    const items = fs.readdirSync(dir);
    for (const item of items) {
        const fullPath = path.join(dir, item);
        if (item === 'Electron.app' && fs.statSync(fullPath).isDirectory()) {
            return fullPath;
        }
        if (fs.statSync(fullPath).isDirectory()) {
            const found = findElectronApp(fullPath);
            if (found) return found;
        }
    }
    return null;
}

/**
 * Copy directory recursively
 */
function copyDirSync(src, dest) {
    fs.mkdirSync(dest, { recursive: true });
    const entries = fs.readdirSync(src, { withFileTypes: true });
    
    for (const entry of entries) {
        const srcPath = path.join(src, entry.name);
        const destPath = path.join(dest, entry.name);
        
        if (entry.isDirectory()) {
            copyDirSync(srcPath, destPath);
        } else {
            fs.copyFileSync(srcPath, destPath);
        }
    }
}

/**
 * Create the app bundle with the provided files.
 */
function createAppBundle(tempDir, appName, mainJsPath, indexHtmlPath, assetPaths) {
    // Find the Electron.app in the extracted files
    const electronAppPath = findElectronApp(tempDir);
    
    if (!electronAppPath) {
        throw new Error("Could not find Electron.app in extracted files");
    }
    
    // Rename Electron.app to the app name
    const appBundlePath = path.join(tempDir, `${appName}.app`);
    fs.renameSync(electronAppPath, appBundlePath);
    
    // Create the app directory inside the bundle
    const appResourcesDir = path.join(appBundlePath, 'Contents', 'Resources', 'app');
    fs.mkdirSync(appResourcesDir, { recursive: true });
    
    // Copy main.js
    fs.copyFileSync(mainJsPath, path.join(appResourcesDir, 'main.js'));
    
    // Copy index.html
    fs.copyFileSync(indexHtmlPath, path.join(appResourcesDir, 'index.html'));
    
    // Copy additional assets
    for (const assetPath of assetPaths) {
        const stat = fs.statSync(assetPath);
        if (stat.isFile()) {
            fs.copyFileSync(assetPath, path.join(appResourcesDir, path.basename(assetPath)));
        } else if (stat.isDirectory()) {
            // Copy directory contents
            const destDir = path.join(appResourcesDir, path.basename(assetPath));
            copyDirSync(assetPath, destDir);
        }
    }
    
    // Create package.json
    const packageJson = {
        name: appName.toLowerCase().replace(/\s+/g, '-'),
        version: "1.0.0",
        main: "main.js"
    };
    fs.writeFileSync(
        path.join(appResourcesDir, 'package.json'),
        JSON.stringify(packageJson, null, 2)
    );
    
    // Update Info.plist to use the app name
    const infoPlistPath = path.join(appBundlePath, 'Contents', 'Info.plist');
    if (fs.existsSync(infoPlistPath)) {
        let content = fs.readFileSync(infoPlistPath, 'utf8');
        // Replace Electron with app name in key places
        content = content.replace('<string>Electron</string>', `<string>${appName}</string>`);
        content = content.replace('<string>com.github.Electron</string>', `<string>com.gazel.${appName}</string>`);
        fs.writeFileSync(infoPlistPath, content);
    }
    
    return appBundlePath;
}

/**
 * Set execute permissions on a file
 */
function setExecutePermissions(filePath) {
    if (fs.existsSync(filePath)) {
        fs.chmodSync(filePath, 0o755);
        console.log(`Set execute permissions on ${filePath}`);
    }
}

/**
 * Create a tar archive of the app bundle.
 */
async function createTarArchive(appBundlePath, outputTarPath) {
    console.log(`Creating tar archive at ${outputTarPath}`);
    
    // First, fix permissions on the executable
    const executablePath = path.join(appBundlePath, 'Contents', 'MacOS', 'Electron');
    setExecutePermissions(executablePath);
    
    // Fix permissions on the main framework
    const frameworkPath = path.join(
        appBundlePath, 'Contents', 'Frameworks',
        'Electron Framework.framework', 'Versions', 'A',
        'Electron Framework'
    );
    setExecutePermissions(frameworkPath);
    
    // Also fix permissions on helper executables
    const helpersDir = path.join(
        appBundlePath, 'Contents', 'Frameworks',
        'Electron Framework.framework', 'Versions', 'A', 'Helpers'
    );
    if (fs.existsSync(helpersDir)) {
        const helpers = fs.readdirSync(helpersDir);
        for (const helper of helpers) {
            const helperPath = path.join(helpersDir, helper);
            const stat = fs.lstatSync(helperPath);
            if (stat.isFile() && !stat.isSymbolicLink()) {
                setExecutePermissions(helperPath);
            }
        }
    }
    
    // Fix permissions on libraries
    const libsDir = path.join(
        appBundlePath, 'Contents', 'Frameworks',
        'Electron Framework.framework', 'Versions', 'A', 'Libraries'
    );
    if (fs.existsSync(libsDir)) {
        const libs = fs.readdirSync(libsDir);
        for (const lib of libs) {
            if (lib.endsWith('.dylib')) {
                const libPath = path.join(libsDir, lib);
                const stat = fs.lstatSync(libPath);
                if (stat.isFile() && !stat.isSymbolicLink()) {
                    setExecutePermissions(libPath);
                }
            }
        }
    }
    
    // Create tar with proper symlink handling
    const appName = path.basename(appBundlePath);
    
    // Change to parent directory to create relative paths in tar
    const parentDir = path.dirname(appBundlePath);
    
    await tar.create(
        {
            file: outputTarPath,
            cwd: parentDir,
            preservePaths: false,
            portable: true,
            follow: false, // Don't follow symlinks, preserve them
        },
        [appName]
    );
    
    console.log(`Created tar archive: ${outputTarPath}`);
}

/**
 * Create a temporary directory
 */
function createTempDir() {
    return fs.mkdtempSync(path.join(os.tmpdir(), 'electron-bundler-'));
}

/**
 * Remove directory recursively
 */
function removeDirSync(dir) {
    if (fs.existsSync(dir)) {
        fs.rmSync(dir, { recursive: true, force: true });
    }
}

/**
 * Main function
 */
async function main() {
    const args = process.argv.slice(2);
    
    if (args.length < 5) {
        console.error("Usage: bundler.js <output_tar> <app_name> <main_js> <index_html> <electron_zip> [assets...]");
        process.exit(1);
    }
    
    const [outputTar, appName, mainJsPath, indexHtmlPath, electronZipPath, ...assetPaths] = args;
    
    console.log(`Building Electron app: ${appName}`);
    console.log(`Output: ${outputTar}`);
    console.log(`Main JS: ${mainJsPath}`);
    console.log(`Index HTML: ${indexHtmlPath}`);
    console.log(`Electron ZIP: ${electronZipPath}`);
    console.log(`Assets: ${assetPaths}`);
    
    // Create a temporary directory for extraction
    const tempDir = createTempDir();
    
    try {
        // Extract Electron
        extractElectron(electronZipPath, tempDir);
        
        // Create the app bundle
        const appBundlePath = createAppBundle(
            tempDir, appName, mainJsPath, indexHtmlPath, assetPaths
        );
        
        // Create the tar archive
        await createTarArchive(appBundlePath, outputTar);
        
        console.log(`Successfully created ${appName} bundle`);
    } finally {
        // Clean up temp directory
        removeDirSync(tempDir);
    }
}

// Run main function
main().catch(error => {
    console.error('Error:', error.message);
    process.exit(1);
});
